### PR TITLE
Create get_audio_devices list from set.

### LINKED ISSUE #174: sounddevice present multiple times in the dropdown list
The speaker/microphone sounddevice is always present multiple times in the respective dropdown list.
It doesn't seem to make a difference which one you choose, but it is irritating.

Problem occured on Windows 10 and Windows 11:

image

### PR DESCRIPTION
**Pull request by kronenpj**
This eliminates duplicates without a lot of gyrations.

We should consider a release from this branch to see if this fix addresses @DK5SM's issue in #174 unless he can build from source to verify.

### CODE REVIEW

**Comment by kronenpj** [comment]
Unfortunately this fix emits:
`2022-06-25 20:04:20 [error    ] [DMN] update_audio_devices: Exception gathering audio devices: e=TypeError("unhashable type: 'dict'")` so a method heavier in gyration will need to be used.

**Review by DJ2LS — APPROVED** [review]
Interesting approach. This could work I think. 
We could publish this now as a release so we are staying at one Update branch, So we are not confusing testers because of different release origins. This might become easier when moving to a first beta release. 
Early testers can use alpha releases then. All others should stay at beta or stable.

**Comment by DJ2LS** [comment]
@kronenpj I'm not sure, but can we merge this? Long pause

**Review by DJ2LS — APPROVED** [review]

**Comment by dk5sm** [comment]
Paul,

I am sorry to say, but this seems not to work either (started FreeDATA.exe from /win-unpacked folder):
image

**Comment by DJ2LS** [comment]
I'm not sure if the build process is working 100% as expected, but it might be possible this is a gui problem. So we should do some filtering on gui side as well. 

As soon as @dk5sm posted the select part of the html code, we can determine how we can do filtering.

**Comment by dk5sm** [comment]
image
image

**Comment by DJ2LS** [comment]
https://github.com/DJ2LS/FreeDATA/blob/690f400a81d975d470c25f6cd89b145bdf7d7c6c/gui/preload-main.js#L1380


Area of interest

**Comment by DJ2LS** [comment]
I suggest we should also add a crc check for determining if it's really a unique audio device like done with serial devices. So we can also check if it's a  client specific behavior


```

          for port, desc, hwid in ports:
                    # calculate hex of hwid if we have unique names
                    crc_hwid = self.crc_algorithm(bytes(hwid, encoding="utf-8"))
                    crc_hwid = crc_hwid.to_bytes(2, byteorder="big")
                    crc_hwid = crc_hwid.hex()
                    description = f"{desc} [{crc_hwid}]"
                    serial_devices.append(
                        {"port": str(port), "description": str(description)}
                    )

```

**Comment by kronenpj** [comment]
Gah! The code checkout portion of the build script tied the build to the `main` branch rather than the one that caused the action to start. Output from this run will have the fixes from this pull request:
https://github.com/DJ2LS/FreeDATA/actions/runs/3086117718

File: tnc/audio.py [hunk 1]
```diff
@@ -49,21 +64,33 @@ def fetch_audio_devices(input_devices, output_devices):
...
+            continue
         except Exception as err:
             print(err)
             max_input_channels = 0
             max_output_channels = 0
             name = ""
 
         if max_input_channels > 0:
-            input_devices.append({"id": index, "name": name})
+            input_devs.add(device_crc(name))
```
Comment (DJ2LS) [inline tnc/audio.py:89]: We should also include additional information like channels, supported sample rate and so on ( maybe all information which is stored in "devices"), so we can be sure we have a unique checksum, even if the names are equal ( happens with Icom ic705 for example. Same usb device name but different settings)

**Review by DJ2LS — APPROVED** [review]

File: tnc/audio.py (hunk 1 above)
Comment (kronenpj) [reply tnc/audio.py]: Done. All data in the `device` structure is included in the checksum.

**Review by DJ2LS — APPROVED** [review]

**Comment by DJ2LS** [comment]
Nice, @kronenpj ! Let's see if these steps help us in debugging.

We can merge it in my opinion, then let's see if the crc is unique enough.
I'm really interested what's the origin of the topics problem 🤔

**Pull request merged**